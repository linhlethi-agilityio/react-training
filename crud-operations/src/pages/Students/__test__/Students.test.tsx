import { render, fireEvent, waitFor } from '@testing-library/react';

import StudentsPage from '..';

// Mocking dependencies
jest.mock('@hooks', () => ({
  useToastCustom: jest.fn(),
  useStudents: () => ({
    students: [
      {
        enrollNumber: 54547,
        name: 'name 8',
        avatarUrl: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/956.jpg',
        phone: '7305477760',
        email: 'Carson31@gmail.com',
        dateOfAdmission: 950288400000,
        id: '0',
      },
    ],
    isLoading: false,
    deleteStudent: jest.fn(),
  }),
}));

jest.mock('jose', () => ({}));

jest.mock('@constants', () => ({
  ENVS: {
    VITE_API_ENDPOINT: 'import.meta.env.VITE_API_ENDPOINT',
    VITE_SECRET_KEY: 'import.meta.env.VITE_SECRET_KEY',
  },
  ERROR_MESSAGES: {
    FIELD_REQUIRED: 'This field is required',
    MIN_PASSWORD_LENGTH: jest.fn(),
  },
  REGEX_PATTERN: {
    EMAIL:
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
    NUMBER: /^[0-9]{10}$/,
    NAME: /^[a-zA-Z ]{1,40}$/,
  },
}));

describe('StudentsPage Component', () => {
  test('renders correctly', () => {
    const { container } = render(<StudentsPage keyword="" />);

    expect(container).toMatchSnapshot();
  });

  test('opens add student modal on button click', () => {
    const { getByText } = render(<StudentsPage keyword="" />);
    fireEvent.click(getByText('Add new student'));

    expect(getByText('Add Student')).toBeInTheDocument(); // Assuming there's a heading with text 'Add Student' in the modal
  });

  test('call handleEditStudent func when on button click', () => {
    const { getByTestId, getByText } = render(<StudentsPage keyword="" />);
    fireEvent.click(getByTestId('edit-button'));

    expect(getByText('Update Student')).toBeInTheDocument(); // Assuming there's a heading with text 'Add Student' in the modal
  });

  test('opens confirm delete modal on delete button click', async () => {
    const { getByTestId, getByText } = render(<StudentsPage keyword="" />);

    fireEvent.click(getByTestId('delete-button'));

    await waitFor(() => {
      expect(getByText('Delete Student')).toBeInTheDocument(); // Assuming there's a heading with text 'Delete Student' in the modal
      expect(getByText('Are you sure you would like to delete student')).toBeInTheDocument();
    });
  });
});
