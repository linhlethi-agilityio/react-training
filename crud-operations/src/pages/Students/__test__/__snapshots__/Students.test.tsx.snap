// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentsPage Component renders correctly 1`] = `
<div>
  <div
    class="css-jj9sx4"
  >
    <div
      class="css-12teqml"
    >
      <h2
        class="chakra-heading css-0"
      >
        Students List
      </h2>
      <div
        class="css-nn803e"
      >
        <button
          aria-label="sort"
          class="chakra-button css-4xx2wk"
          type="button"
        >
          <svg
            fill="#000000"
            height="20"
            id="Capa_1"
            version="1.1"
            viewBox="0 0 562.392 562.391"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <g>
                <path
                  d="M123.89,262.141h314.604c19.027,0,17.467-31.347,15.496-47.039c-0.605-4.841-3.636-11.971-6.438-15.967L303.965,16.533
 c-12.577-22.044-32.968-22.044-45.551,0L114.845,199.111c-2.803,3.996-5.832,11.126-6.438,15.967
 C106.43,230.776,104.863,262.141,123.89,262.141z"
                  fill="#fff"
                  stroke="#FEAF00"
                  stroke-width="20px"
                />
                <path
                  d="M114.845,363.274l143.569,182.584c12.577,22.044,32.968,22.044,45.551,0l143.587-182.609
 c2.804-3.996,5.826-11.119,6.438-15.967c1.971-15.691,3.531-47.038-15.496-47.038H123.89c-19.027,0-17.46,31.365-15.483,47.062
 C109.019,352.147,112.042,359.277,114.845,363.274z"
                  fill="#fff"
                  stroke="#FEAF00"
                  stroke-width="20px"
                />
              </g>
            </g>
          </svg>
        </button>
        <button
          aria-label="add-student"
          class="chakra-button css-4xx2wk"
          type="button"
        >
          Add new student
        </button>
      </div>
    </div>
    <div
      class="css-0"
    >
      <table
        class="chakra-table css-0"
      >
        <thead
          class="css-0"
        >
          <tr
            class="css-0"
          >
            <th
              class="css-tln5s8"
            />
            <th
              class="css-tln5s8"
            >
              Name
            </th>
            <th
              class="css-tln5s8"
            >
              Email
            </th>
            <th
              class="css-tln5s8"
            >
              Phone
            </th>
            <th
              class="css-tln5s8"
            >
              Enroll Number
            </th>
            <th
              class="css-tln5s8"
            >
              Date of admission
            </th>
            <th
              class="css-tln5s8"
            />
          </tr>
        </thead>
        <tbody
          class="css-0"
        >
          <tr
            class="css-1w3yp13"
          >
            <td
              class="css-fsw37u"
            >
              <span
                class="chakra-avatar css-1v9juh0"
              >
                <svg
                  aria-label=" avatar"
                  class="chakra-avatar__svg css-16ite8i"
                  role="img"
                  viewBox="0 0 128 128"
                >
                  <path
                    d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                    fill="currentColor"
                  />
                  <path
                    d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </td>
            <td
              class="css-fsw37u"
            >
              name 8
            </td>
            <td
              class="css-fsw37u"
            >
              <p
                class="chakra-text css-jby5hl"
              >
                Carson31@gmail.com
              </p>
            </td>
            <td
              class="css-fsw37u"
            >
              7305477760
            </td>
            <td
              class="css-fsw37u"
            >
              54547
            </td>
            <td
              class="css-fsw37u"
            >
              11-Feb, 2000
            </td>
            <td
              class="css-fsw37u"
            >
              <div
                class="chakra-stack css-85kjo3"
              >
                <button
                  aria-label="edit"
                  class="chakra-button css-4xx2wk"
                  data-testid="edit-button"
                  type="button"
                >
                  <svg
                    class="chakra-icon css-13otjrl"
                    focusable="false"
                    height="19"
                    viewBox="0 0 19 19"
                    width="19"
                  >
                    <path
                      d="M18.303 2.088 16.911.696A2.366 2.366 0 0 0 15.232 0c-.608 0-1.216.232-1.68.695L.476 13.772l-.471 4.239a.89.89 0 0 0 .983.983l4.235-.468 13.08-13.08a2.374 2.374 0 0 0 0-3.358ZM4.678 17.392l-3.452.383.384-3.457 9.793-9.793 3.072 3.071-9.797 9.796ZM17.464 4.607l-2.15 2.15-3.071-3.072 2.15-2.15a1.18 1.18 0 0 1 .84-.347c.316 0 .614.123.839.347l1.392 1.392a1.19 1.19 0 0 1 0 1.68Z"
                      fill="#FEAF00"
                    />
                  </svg>
                </button>
                <button
                  aria-label="delete"
                  class="chakra-button css-4xx2wk"
                  data-testid="delete-button"
                  type="button"
                >
                  <svg
                    class="chakra-icon css-13otjrl"
                    focusable="false"
                    height="18"
                    viewBox="0 0 16 18"
                    width="16"
                  >
                    <path
                      d="M.286 2.25H4L5.2.675c.16-.21.367-.38.605-.497.238-.117.5-.178.766-.178H9.43c.266 0 .528.061.766.178.238.117.445.287.605.497L12 2.25h3.714c.076 0 .149.03.202.082a.279.279 0 0 1 .084.2v.562c0 .074-.03.146-.084.199a.288.288 0 0 1-.202.082h-.675l-1.185 13.089c-.039.42-.235.81-.551 1.094a1.73 1.73 0 0 1-1.157.442H3.854a1.73 1.73 0 0 1-1.157-.442 1.678 1.678 0 0 1-.55-1.094L.96 3.375H.286a.288.288 0 0 1-.202-.082.279.279 0 0 1-.084-.2v-.562c0-.074.03-.146.084-.199a.288.288 0 0 1 .202-.082Zm9.6-.9a.573.573 0 0 0-.457-.225H6.57a.583.583 0 0 0-.457.225l-.685.9h5.142l-.685-.9Zm-6.6 15.012c.011.14.076.27.182.366a.573.573 0 0 0 .386.147h8.292c.143 0 .281-.052.386-.147a.556.556 0 0 0 .182-.366l1.179-12.987H2.107l1.179 12.987Z"
                      fill="#FEAF00"
                    />
                  </svg>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
